utils/src/treev2/quickPickWizard/ContextValueQuickPickStep.ts

```diff
@@ -0,0 +1,53 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { TreeItem } from 'vscode';
+import * as types from '../../../index';
+import { parseContextValue } from '../../utils/contextUtils';
+import { PickFilter } from './common/PickFilter';
+import { GenericQuickPickOptions, GenericQuickPickStep } from './GenericQuickPickStep';
+
+export interface ContextValueFilterQuickPickOptions extends GenericQuickPickOptions {
+    contextValueFilter: types.ContextValueFilter;
+}
+
+export class ContextValueQuickPickStep<TContext extends types.QuickPickWizardContext, TOptions extends ContextValueFilterQuickPickOptions> extends GenericQuickPickStep<TContext, TOptions> {
+    protected readonly pickFilter: PickFilter = new ContextValuePickFilter(this.pickOptions);
+}
+
+class ContextValuePickFilter implements PickFilter {
+    constructor(private readonly pickOptions: ContextValueFilterQuickPickOptions) { }
+
+    isDirectPick(node: TreeItem): boolean {
+        const includeOption = this.pickOptions.contextValueFilter.include;
+        const excludeOption = this.pickOptions.contextValueFilter.exclude;
+
+        const includeArray: (string | RegExp)[] = Array.isArray(includeOption) ? includeOption : [includeOption];
+        const excludeArray: (string | RegExp)[] = excludeOption ?
+            (Array.isArray(excludeOption) ? excludeOption : [excludeOption]) :
+            [];
+
+        const nodeContextValues: string[] = parseContextValue(node.contextValue);
+
+        return includeArray.some(i => this.matchesSingleFilter(i, nodeContextValues)) &&
+            !excludeArray.some(e => this.matchesSingleFilter(e, nodeContextValues));
+    }
+
+    isIndirectPick(node: TreeItem): boolean {
+        // `TreeItemCollapsibleState.None` and `undefined` are both falsy, and indicate that a `TreeItem` cannot have children--and therefore, cannot be an indirect pick
+        return !node.collapsibleState;

```

Comment:
This step is used by both the compatibility experience, and the non-compatibility experiences. So `TreeItem` won't necessarily be an `AzExtTreeItem`.

> I know that we want to use vscode's implementation of TreeItem but I feel uneasy relying on a property that vscode controls and changes whenever it sees fit.

I don't think VS Code changes this property. For example, we define it only as a `getter` in `AzExtParentTreeItem`.

Reply:
> Does it ever come in as undefined because the collapsibleState hadn't been set yet? I remember you were playing with this quite a bit.

If you don't set `collapsibleState`, then VS Code treats it as `TreeItemCollapsibleState.None` which has a value of 0. It can do this since `undefined` and `0` are falsy.